src/Features/Core/Portable/Diagnostics/Analyzers/NamingStyles/NamingStyleDiagnosticAnalyzerBase.cs

```diff
@@ -0,0 +1,125 @@
+                // any change. The overall system would then only deserialize & allocate when 
+                // actually necessary.
+                var viewModel = Deserialize(currentValue, typeof(SerializableNamingStylePreferencesInfo)) as SerializableNamingStylePreferencesInfo;
+                var preferencesInfo = viewModel.GetPreferencesInfo();
+                context.RegisterSymbolAction(
+                    symbolContext => SymbolAction(symbolContext, preferencesInfo),
+                    _symbolKinds);
+            }
+        }
+
+        private void SymbolAction(SymbolAnalysisContext context, NamingStylePreferencesInfo preferences)
+        {
+            NamingRule applicableRule;
+            if (preferences.TryGetApplicableRule(context.Symbol, out applicableRule))
+            {
+                string failureReason;
+                if (applicableRule.EnforcementLevel != DiagnosticSeverity.Hidden && 
+                    !applicableRule.IsNameCompliant(context.Symbol.Name, out failureReason))
+                {
+                    var descriptor = new DiagnosticDescriptor(IDEDiagnosticIds.NamingRuleId,
+                         "Title", // Todo: strings
+                         string.Format(FeaturesResources.NamingViolationDescription, applicableRule.Title, failureReason),
+                         DiagnosticCategory.Style,
+                         applicableRule.EnforcementLevel,
+                         isEnabledByDefault: true);
+
+                    var builder = ImmutableDictionary.CreateBuilder<string, string>();
+                    builder[nameof(NamingStyle)] = Serialize(applicableRule.NamingStyle);
+                    builder["OptionName"] = nameof(SimplificationOptions.NamingPreferences); // TODO: doc this
+                    builder["OptionLanguage"] = LanguageNames.CSharp;
+                    context.ReportDiagnostic(Diagnostic.Create(descriptor, context.Symbol.Locations.First(), builder.ToImmutable()));
+                }
+            }
+        }
+
+        public static string Serialize(object obj)
+        {
+            using (MemoryStream memoryStream = new MemoryStream())
+            using (StreamReader reader = new StreamReader(memoryStream))
+            {
+                DataContractSerializer serializer = new DataContractSerializer(obj.GetType());
+                serializer.WriteObject(memoryStream, obj);
+                memoryStream.Position = 0;
+                return reader.ReadToEnd();
+            }
+        }
+
+        public static object Deserialize(string xml, Type toType)
+        {
+            try
+            {
+                using (Stream stream = new MemoryStream())
+                {
+                    byte[] data = System.Text.Encoding.UTF8.GetBytes(xml);
+                    stream.Write(data, 0, data.Length);
+                    stream.Position = 0;
+                    DataContractSerializer deserializer = new DataContractSerializer(toType);
```

Comment:
is DataContractSerializer safe across versions of VS.  are these things you're serializing/deserializing ephemeral, or persisted anywhere?